### PR TITLE
[ONNX] Single function for torch.onnx.export and torch.onnx.dynamo_export

### PR DESCRIPTION
**Pull request by titaiwangms**
Add `dynamo: bool = True` as a switch in `torch.onnx.export` to provide users an option to try `torch.onnx.dynamo_export`.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -508,6 +511,36 @@ def forward(self, x):
             All errors are subclasses of :class:`errors.OnnxExporterError`.
     """
 
+    if dynamo:
+        # Unsupported parameters for dynamo export
+        # TODO: These are not supported AT THE TIME
+        warnings.warn(
+            "f, export_params, verbose, training, input_names, output_names, operator_export_type, opset_version, \
+            do_constant_folding, keep_initializers_as_inputs, custom_opsets, export_modules_as_functions, and \
+            autograd_inlining are not supported for dynamo export at the moment."
```
Comment (justinchuby) [inline torch/onnx/utils.py:520]:
```suggestion
            "f, export_params, verbose, training, input_names, output_names, operator_export_type, opset_version, "
            "do_constant_folding, keep_initializers_as_inputs, custom_opsets, export_modules_as_functions, and "
            "autograd_inlining are not supported for dynamo export at the moment."
```

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -508,6 +511,36 @@ def forward(self, x):
...
+            warnings.warn(
+                "Specified dynamic axes is not supported for dynamo export at the moment."
+            )
+        # TODO: expose more ExportOptions?
+        export_options = torch.onnx.ExportOptions(dynamic_shapes=dynamic_shapes)
+        return torch.onnx.dynamo_export(
+            model, *args, **kwargs, export_options=export_options
+        )
+
+    assert f is not None, "Exporting to a file is required for non-dynamo exporting."
```
Comment (justinchuby) [inline torch/onnx/utils.py:543]: Raise a value error instead

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -508,6 +511,36 @@ def forward(self, x):
...
+        if dynamic_axes is None or not isinstance(dynamic_axes, dict):
+            dynamic_shapes = False
+        else:
+            dynamic_shapes = True
+            warnings.warn(
+                "Specified dynamic axes is not supported for dynamo export at the moment."
+            )
+        # TODO: expose more ExportOptions?
+        export_options = torch.onnx.ExportOptions(dynamic_shapes=dynamic_shapes)
+        return torch.onnx.dynamo_export(
```
Comment (justinchuby) [inline torch/onnx/utils.py:539]: Handle when f is provided?

**Review by justinchuby — APPROVED** [review]
LGTM thanks!

File: torch/onnx/utils.py (hunk 3 above)
Comment (titaiwangms) [reply torch/onnx/utils.py:539]: ONNXProgram.save to f?

Comment (justinchuby) [reply torch/onnx/utils.py:539]: That sounds good

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -508,6 +510,43 @@ def forward(self, x):
...
+            warnings.warn(
+                "Specified dynamic axes is not supported for dynamo export at the moment."
+            )
+        # TODO: expose more ExportOptions?
+        export_options = torch.onnx.ExportOptions(dynamic_shapes=dynamic_shapes)
+        onnx_program = torch.onnx.dynamo_export(
+            model, *args, **kwargs, export_options=export_options
+        )
+        if f is not None:
+            onnx_program.save(f)
```
Comment (titaiwangms) [inline torch/onnx/utils.py:542]: add a test

File: test/onnx/dynamo/test_exporter_api.py [hunk 5]
```diff
@@ -221,5 +228,94 @@ def test_serialize_succeeds_when_model_greater_than_2gb(self):
...
+    def test_args_normalization_with_no_kwargs(self):
+        onnx_program_from_new_exporter = torch.onnx.dynamo_export(
+            SampleModelTwoInputs(), torch.randn(1, 1, 2), torch.randn(1, 1, 2)
+        )
+        onnx_program_from_old_exporter = torch.onnx.export(
+            SampleModelTwoInputs(),
+            (torch.randn(1, 1, 2), torch.randn(1, 1, 2)),
+            dynamo=True,
+        )
+        self.assertEqual(
```
Comment (gramalingam) [inline test/onnx/dynamo/test_exporter_api.py:241]: Can you say something about how this equality assertion works? Just for my own education, wondering what kind of infrastructure exists for proto-object equality check.

Comment (titaiwangms) [reply test/onnx/dynamo/test_exporter_api.py:241]: That's a good point. I didn't verify that. This might be invalid. I will verify this.

Comment (justinchuby) [reply test/onnx/dynamo/test_exporter_api.py:241]: There is some internal equality check according to https://proto-plus-python.readthedocs.io/en/stable/_modules/proto/message.html. I think it should work when two messages are _exactly_ the same.

Comment (titaiwangms) [reply test/onnx/dynamo/test_exporter_api.py:241]: It does catch the difference between two protobuf if there is any. Here is their code: https://github.com/pytorch/pytorch/blob/22964d1007d1d87964df1db609da53b35b00316f/torch/testing/_internal/common_utils.py#L3522

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -508,6 +510,43 @@ def forward(self, x):
...
+        # Unsupported parameters for dynamo export
+        # TODO: These are not supported AT THE TIME
+        warnings.warn(
+            "f, export_params, verbose, training, input_names, output_names, operator_export_type, opset_version, "
+            "do_constant_folding, keep_initializers_as_inputs, custom_opsets, export_modules_as_functions, and "
+            "autograd_inlining are not supported for dynamo export at the moment."
+        )
+        # TODO: check args normalization
+        args = _decide_input_format(model, args)
+        kwargs = {}
```
Comment (justinchuby) [inline torch/onnx/utils.py:523]: In a follow up PR, I would add an kwarg parameter to the api and encourage users to use that.

File: torch/onnx/utils.py [hunk 7]
```diff
@@ -186,11 +186,10 @@ def exporter_context(model, mode: _C_onnx.TrainingMode, verbose: bool):
         yield (mode_ctx, apex_ctx, log_ctx, diagnostic_ctx)
 
 
-@_beartype.beartype
 def export(
     model: Union[torch.nn.Module, torch.jit.ScriptModule, torch.jit.ScriptFunction],
     args: Union[Tuple[Any, ...], torch.Tensor],
-    f: Union[str, io.BytesIO],
+    f: Optional[Union[str, io.BytesIO]] = None,
     export_params: bool = True,
```
Comment (justinchuby) [inline torch/onnx/utils.py:193]:
```suggestion
    *,
    export_params: bool = True,
```

as a follow up

File: torch/onnx/utils.py [hunk 8]
```diff
@@ -206,7 +205,8 @@ def export(
     custom_opsets: Optional[Mapping[str, int]] = None,
     export_modules_as_functions: Union[bool, Collection[Type[torch.nn.Module]]] = False,
     autograd_inlining: Optional[bool] = True,
```
Comment (justinchuby) [inline torch/onnx/utils.py:207]:
```suggestion
    autograd_inlining: Optional[bool] = True,
    kwargs: Mapping[str, Any] = None
```

As a follow up

File: torch/onnx/utils.py (hunk 6 above)
Comment (titaiwangms) [reply torch/onnx/utils.py:523]: torch.onnx.export has its own way to support kwargs. Do we want to add another way of supporting kwargs?
https://github.com/pytorch/pytorch/blob/4adee71155bec4e419bac32be2cbc1763bc6c98f/torch/onnx/utils.py#L240-L281

**Pull request closed without merging**